<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [1]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0

  Attempting uninstall: yfinance
    Found existing installation: yfinance 0.2.4
    Uninstalling yfinance-0.2.4:
      Successfully uninstalled yfinance-0.2.4

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████

In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [4]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [5]:
tesla = yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [9]:
tesla_data = tesla.history(period="max")


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [8]:

tesla_data.reset_index(inplace=True)
print(tesla_data.head())


        Date      Open      High       Low     Close     Volume  Dividends  \
0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  


## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [10]:
import requests

# URL of the webpage to download
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm'

# Send an HTTP GET request to the URL
html_data = requests.get(url)

# Check if the request was successful (status code 200)
if html_data.status_code == 200:
    # Save the text of the response as html_data
    html_data = html_data.text  # Corrected this line
    print("Webpage downloaded successfully.")
else:
    print("Failed to download the webpage. Status code:", html_data.status_code)


Webpage downloaded successfully.


Parse the html data using `beautiful_soup`.


In [11]:
soup = BeautifulSoup(html_data, 'html.parser')

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [13]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# URL of the webpage to download
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm'

# Send an HTTP GET request to the URL
html_data = requests.get(url)

# Check if the request was successful (status code 200)
if html_data.status_code == 200:
    # Parse the HTML content with BeautifulSoup
    soup = BeautifulSoup(html_data.text, 'html.parser')

    # Find the table containing Tesla's revenue data
    table = soup.find('table')

    # Use Pandas to read the HTML table into a DataFrame
    tesla_revenue = pd.read_html(str(table))[0]

    # Rename columns to 'Date' and 'Revenue'
    tesla_revenue.columns = ['Date', 'Revenue']

    # Display the first few rows of the DataFrame
    print(tesla_revenue.head())
else:
    print("Failed to download the webpage. Status code:", html_data.status_code)

In [83]:
import matplotlib.pyplot as plt
import pandas as pd
import requests
from bs4 import BeautifulSoup

# Define the URL where the revenue data is hosted
revenue_url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm'

# Send an HTTP GET request to the URL
response = requests.get(revenue_url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Parse the HTML content using BeautifulSoup
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # Find the table that contains the revenue data (assuming it's the first table)
    table = soup.find_all('table')[0]
    
    # Extract the table data and create the tesla_revenue DataFrame
    data = []
    for row in table.find_all('tr'):
        cols = row.find_all('td')
        if len(cols) == 2:  # Assuming there are two columns: Date and Revenue
            date = cols[0].get_text()
            revenue = cols[1].get_text()
            data.append([date, revenue])
    
    # Create the tesla_revenue DataFrame
    tesla_revenue = pd.DataFrame(data, columns=['Date', 'Revenue'])
    
    # Preprocess the revenue data
    tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace('[\$,]', '', regex=True)
    tesla_revenue.dropna(inplace=True)
    tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]
    
    # Fetch Tesla stock data using yfinance
    tesla_data = yf.download('TSLA', start='2020-01-01', end='2021-01-01')  # Adjust the date range as needed

    def make_graph(stock_data, revenue_data, title, save_path=None):
        # Plot the stock data
        plt.figure(figsize=(12, 6))
        plt.plot(stock_data.index, stock_data['Close'], label='Stock Price', color='b')
        plt.xlabel('Date')
        plt.ylabel('Stock Price')
        
        # Add revenue data to the graph
        plt.bar(pd.to_datetime(revenue_data['Date']), revenue_data['Revenue'], label='Revenue', alpha=0.5, color='g')
        plt.ylabel('Revenue')
        
        # Set the title and labels
        plt.title(f'{title} Stock Price and Revenue')
        plt.legend()
        plt.grid(True)
        
        # Save the graph as an image if save_path is provided
        if save_path:
            plt.savefig(save_path)
            print(f"Graph saved as {save_path}")
        else:
            plt.show()

    # Example usage for Tesla:
    make_graph(tesla_data, tesla_revenue, 'Tesla', save_path='tesla_graph.png')
    
else:
    print("Failed to fetch revenue data. Status code:", response.status_code)

[*********************100%***********************]  1 of 1 completed
Graph saved as tesla_graph.png


Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [77]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace('[\$,]', '', regex=True)

Execute the following lines to remove an null or empty strings in the Revenue column.


In [78]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace('[\$,]', '', regex=True)
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [16]:
print(tesla_revenue.tail())

    Date Revenue
8   2013    2013
9   2012     413
10  2011     204
11  2010     117
12  2009     112


## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [24]:
GameStop = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [26]:
gme_data = GameStop.history(period="max")


**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [27]:
gme_data.reset_index(inplace=True)
print(gme_data.head())

        Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620128  1.693350  1.603296  1.691666  76216000        0.0   
1 2002-02-14  1.712707  1.716074  1.670626  1.683250  11021600        0.0   
2 2002-02-15  1.683250  1.687458  1.658001  1.674834   8389600        0.0   
3 2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615920  1.662210  1.603296  1.662210   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  


## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [28]:
import requests

# URL of the webpage to download (enclosed in quotes)
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html'

# Send an HTTP GET request to the URL
html_data = requests.get(url)

# Check if the request was successful (status code 200)
if html_data.status_code == 200:
    # Save the text of the response as html_data.text (corrected this line)
    html_data = html_data.text
    print("Webpage downloaded successfully.")
else:
    print("Failed to download the webpage. Status code:", html_data.status_code)


Webpage downloaded successfully.


Parse the html data using `beautiful_soup`.


In [29]:
soup = BeautifulSoup(html_data, 'html.parser')

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [81]:
import pandas as pd

# URL of the webpage with the table
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html'

tables = pd.read_html(url)

gme_revenue = tables[0]

gme_revenue.columns = ['Date', 'Revenue']

print(gme_revenue.head())



   Date Revenue
0  2020  $6,466
1  2019  $8,285
2  2018  $8,547
3  2017  $7,965
4  2016  $9,364


Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [33]:
print(gme_revenue.tail())

    Date Revenue
11  2009  $8,806
12  2008  $7,094
13  2007  $5,319
14  2006  $3,092
15  2005  $1,843


## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [68]:
import matplotlib
matplotlib.use('agg')

In [69]:
pip install --upgrade matplotlib pandas

Note: you may need to restart the kernel to use updated packages.


In [87]:
import matplotlib.pyplot as plt
import pandas as pd
import requests
from bs4 import BeautifulSoup
import yfinance as yf

# Define the URL where the revenue data is hosted
revenue_url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm'

# Send an HTTP GET request to the URL
response = requests.get(revenue_url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Parse the HTML content using BeautifulSoup
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # Find the table that contains the revenue data (assuming it's the first table)
    table = soup.find_all('table')[0]
    
    # Extract the table data and create the tesla_revenue DataFrame
    data = []
    for row in table.find_all('tr'):
        cols = row.find_all('td')
        if len(cols) == 2:  # Assuming there are two columns: Date and Revenue
            date = cols[0].get_text()
            revenue = cols[1].get_text()
            data.append([date, revenue])
    
    # Create the tesla_revenue DataFrame
    tesla_revenue = pd.DataFrame(data, columns=['Date', 'Revenue'])
    
    # Preprocess the revenue data
    tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace('[\$,]', '', regex=True)
    tesla_revenue.dropna(inplace=True)
    tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]
    
    # Fetch Tesla stock data using yfinance
    tesla_data = yf.download('TSLA', start='2020-01-01', end='2021-01-01')  # Adjust the date range as needed

    # Sort both DataFrames by the 'Date' column
    tesla_data.sort_values(by='Date', inplace=True)
    tesla_revenue.sort_values(by='Date', inplace=True)
    
    # Merge the DataFrames based on the 'Date' column
    merged_data = pd.merge(tesla_data, tesla_revenue, on='Date', how='inner')

    def make_graph(stock_data, title, save_path=None):
        # Plot the stock data
        plt.figure(figsize=(12, 6))
        plt.plot(stock_data['Date'], stock_data['Close'], label='Stock Price', color='b')
        plt.xlabel('Date')
        plt.ylabel('Stock Price')
        
        # Add revenue data to the graph
        plt.bar(stock_data['Date'], stock_data['Revenue'], label='Revenue', alpha=0.5, color='g')
        plt.ylabel('Revenue')
        
        # Set the title and labels
        plt.title(title)
        plt.legend()
        plt.grid(True)
        
        # Save the graph as an image if save_path is provided
        if save_path:
            plt.savefig(save_path)
            print(f"Graph saved as {save_path}")
        else:
            plt.show()

    # Example usage for Tesla:
    make_graph(merged_data, 'Tesla Stock Price and Revenue', save_path='tesla_graph.png')


[*********************100%***********************]  1 of 1 completed


ValueError: You are trying to merge on datetime64[ns] and object columns. If you wish to proceed you should use pd.concat

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [63]:
import matplotlib
matplotlib.use('agg')

In [64]:
pip install --upgrade matplotlib pandas


Note: you may need to restart the kernel to use updated packages.


In [65]:
import matplotlib.pyplot as plt

def make_graph(stock_data, revenue_data, title, save_path=None):
    # ... (same plotting code as before)

    # Save the graph as an image if save_path is provided
    if save_path:
        plt.savefig(save_path)
        print(f"Graph saved as {save_path}")
    else:
        plt.show()

# Example usage:
# make_graph(gme_data, gme_revenue, 'GameStop', save_path='gme_graph.png')

In [67]:
make_graph(gme_data, gme_revenue, 'GameStop', save_path='gme_graph.png')

Graph saved as gme_graph.png


<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
